### ISSUE TITLE
Support GS232B Protokoll and CGAntenna RTC-200

### ISSUE THREAD

**Issue by DK4FT** [dh1tw/remoteRotator#28]
Hi Tobias,

great application, really like the easy and clean webinterface.

I am setting up a remote station for our club and would like to use the remoteRotator there.
Our setup consists of a Yaesu G-1000DXC, a CG Antenna RTC-200 USB Interface, remoteRotator is running on a RPi.

I am able to connect to the deivce and turn the antenna, but there is no feedback about the beam direction.
Could you add the device to the supported interfaces, please?

**Comment by dh1tw** [comment]
Hi Alex,

would be interesting to know which protocol the RTC-200 exactly emulates. The Yaesu G232 A and GS232 B and popular ones. You can connect with any terminal to the device. On Windows, Putty should work, on Linux you could use the application "screen".

Please attach the terminal output. 

73 Tobias

**Comment by DK4FT** [comment]
According to the documentation it should emulate GS232-B protocol.

I checked the communication with 
`screen /dev/ttyUSB0 9600`
but there is no output in the terminal window.
I added my user to dialout group, but even then no echo or feedback from the interface...

Next days I will be onsite and check again with the device attached to a windows machine and putty.

73 Alex

**Comment by DK4FT** [comment]
I traced the tty traffic on a windows machine.

Steps I tested:
1. send manual commands with putty on a Win10 machine.
    Result: there is also no response from the rtc-200 device. E.G. If I send the C2 command, no repsonse
2. use the CGantenna provided mini Application
    Result: There is a heartbeat on a LED of the device visible (same as with remoteRotator application)
    in addition the application is sending regular C2 commands and the device responds with the actual Azimuth position of the rotator. Here a snippet from the sniffed tty communication:
`C2
AZ=113  EL=030
C2
AZ=113  EL=030
C2
AZ=113  EL=030
C2
AZ=113  EL=030
C2
AZ=113  EL=030
C2
AZ=113  EL=030
C2
AZ=113  EL=030
C2
AZ=113  EL=030
C2
AZ=113  EL=030
C2
AZ=113  EL=030
C2
AZ=113  EL=030
C2
AZ=113  EL=030`

The only difference between sending data by putty or sending data with the CGantenna app are the controll wires.
When using putty all signals are "off", when using the CGAntenna app all wires are on "low-level" (red)

    
![port_sniffer](https://github.com/dh1tw/remoteRotator/assets/62115413/18e22d1b-ff5f-4abf-9e16-d0541741f458)

**Comment by DK4FT** [comment]
Just read through the GS-232B Manual.
According to the manual C2 is pulling the Elevation Antenna Direction Value.

Is there a second issue and the commands are missused in the CGAntenne device?

**Comment by DK4FT** [comment]
Did a trace using socat in the linux machine 

1. RTC-200 is connected to /dev/ttyUSB1
2. startet `sudo socat -x -v /dev/ttyUSB1,rawer,b9600,crnl PTY,link=/home/AlexFritz/ttyV1,rawer,crnl`
3. startet remoteRotator with interface to /home/AlexFritz/ttyV1:


< 2024/01/17 13:36:36.000368496  length=4 from=0 to=3
 43 32 0d 0a                                      C2..
--
> 2024/01/17 13:36:36.000392172  length=10 from=0 to=9
 41 5a 3d 31 31 30 20 20 45 4c                    AZ=110  EL
--
> 2024/01/17 13:36:36.000408352  length=6 from=10 to=15
 3d 30 33 32 0d 0a                                =032..
--
< 2024/01/17 13:36:37.000368907  length=4 from=4 to=7
 43 32 0d 0a                                      C2..
--
> 2024/01/17 13:36:37.000383883  length=2 from=16 to=17
 41 5a                                            AZ
--
> 2024/01/17 13:36:37.000399948  length=14 from=18 to=31
 3d 31 31 30 20 20 45 4c 3d 30 33 32 0d 0a        =110  EL=032..
--
< 2024/01/17 13:36:38.000368382  length=4 from=8 to=11
 43 32 0d 0a                                      C2..
--
> 2024/01/17 13:36:38.000391750  length=10 from=32 to=41
 41 5a 3d 31 31 30 20 20 45 4c                    AZ=110  EL
--
> 2024/01/17 13:36:38.000407944  length=6 from=42 to=47
 3d 30 33 32 0d 0a                                =032..
--
< 2024/01/17 13:36:39.000368505  length=4 from=12 to=15
 43 32 0d 0a                                      C2..
--
> 2024/01/17 13:36:39.000383562  length=2 from=48 to=49
 41 5a                                            AZ
--
> 2024/01/17 13:36:39.000399562  length=14 from=50 to=63
 3d 31 31 30 20 20 45 4c 3d 30 33 32 0d 0a        =110  EL=032..
--
< 2024/01/17 13:36:40.000368345  length=4 from=16 to=19
 43 32 0d 0a                                      C2..
--
> 2024/01/17 13:36:40.000391380  length=9 from=64 to=72
 41 5a 3d 31 31 30 20 20 45                       AZ=110  E
--
> 2024/01/17 13:36:40.000407620  length=7 from=73 to=79
 4c 3d 30 33 32 0d 0a                             L=032..
--
< 2024/01/17 13:36:41.000368159  length=4 from=20 to=23
 43 32 0d 0a                                      C2..
--
> 2024/01/17 13:36:41.000383356  length=1 from=80 to=80
 41                                               A
--
> 2024/01/17 13:36:41.000399246  length=15 from=81 to=95
 5a 3d 31 31 30 20 20 45 4c 3d 30 33 32 0d 0a     Z=110  EL=032..
--
< 2024/01/17 13:36:42.000368502  length=4 from=24 to=27
 43 32 0d 0a                                      C2..
--
> 2024/01/17 13:36:42.000391205  length=8 from=96 to=103
 41 5a 3d 31 31 30 20 20                          AZ=110
--
> 2024/01/17 13:36:42.000407164  length=8 from=104 to=111
 45 4c 3d 30 33 32 0d 0a                          EL=032..
--
< 2024/01/17 13:36:43.000368826  length=4 from=28 to=31
 43 32 0d 0a                                      C2..
--
> 2024/01/17 13:36:43.000399289  length=15 from=112 to=126
 41 5a 3d 31 31 30 20 20 45 4c 3d 30 33 32 0d     AZ=110  EL=032.
--
> 2024/01/17 13:36:43.000416564  length=1 from=127 to=127
 0a                                               .
--
< 2024/01/17 13:36:44.000369256  length=4 from=32 to=35
 43 32 0d 0a                                      C2..
--
> 2024/01/17 13:36:44.000390908  length=7 from=128 to=134
 41 5a 3d 31 31 30 20                             AZ=110
--
> 2024/01/17 13:36:44.000406869  length=9 from=135 to=143
 20 45 4c 3d 30 33 32 0d 0a                        EL=032..

**Comment by dh1tw** [comment]
Thanks for the traces Alex. I had a look into the source code. So far I only implemented GS232A.

GS232A returns on a `C2` query: 
```
+0030+0090
```

While GS232B returns on a `C2`query:
```
AZ=113  EL=030
```

I will have to implement the GS232B protocol.

**Comment by DK4FT** [comment]
Hi Tobias,

do you have a plan if or estimation when you like to implement the GS323B protocoll?

**Comment by dh1tw** [comment]
I hope I have a version ready for testing this weekend

**Comment by dh1tw** [comment]
Hi Alex,
I have added support for GS232B in the [gs232b](https://github.com/dh1tw/remoteRotator/tree/gs232b) branch.
Could you give it a try with your RTC-200 interface? 
I manually uploaded the binaries from the latest commit (7c1534c65672c0380a9f80fe2c282b59159c7197) here:
https://cloud.dh1tw.de/s/aeNXYqdN8MQLgSr

**Comment by DK4FT** [comment]
Hi Tobias,

thanks for the test-version.
I have downloaded the beta version and gave it a try.

The UI still doesn't display the current heading - no change
![image](https://github.com/dh1tw/remoteRotator/assets/62115413/10be4ed4-e1e0-4312-bc01-6adde3d848e8)


Protocoll from the device is still the same (sure there was no change ;-) )
```
< 2024/01/28 16:42:35.000110045  length=4 from=618 to=621
 43 32 0d 0a                                      C2..
--
> 2024/01/28 16:42:35.000140174  length=15 from=2368 to=2382
 41 5a 3d 30 36 31 20 20 45 4c 3d 30 32 33 0d 0a    AZ=061  EL=023..
```

Are there any debugging switches, which could give you more information/insights?

**Comment by DK4FT** [comment]
Hi Tobias,

i just gave your beta version a try and it seem to work fine:
![image](https://github.com/dh1tw/remoteRotator/assets/62115413/9b79f84f-30d9-4ed5-afd2-1d7c10da4949)

**Comment by dh1tw** [comment]
Excellent. Then I will merge the branch and create a new release (v0.7.0).

**Comment by dh1tw** [comment]
Hi Alex, 

I just published the [v0.7.0 release](/dh1tw/remoteRotator/releases/tag/v0.7.0). In case you find bugs or unexpected behaviour, please re-open or create a new issue.
73 Tobias
